contracts/src/lib/LibGateway.sol

```diff
@@ -339,35 +349,38 @@ library LibGateway {
     }
 
     /// @notice executes a cross message if its destination is the current network, otherwise adds it to the postbox to be propagated further
+    /// This function assumes that the relevant funds have been already minted or burnt
+    /// when the top-down or bottom-up messages have been queued for execution.
+    /// This function is not expected to revert. If a controlled failure happens, a new
+    /// cross-message receipt is propagated for execution to inform the sending contract.
+    /// `Call` cross-messages also trigger receipts if they are successful.
     /// @param arrivingFrom - the immediate subnet from which this message is arriving
     /// @param crossMsg - the cross message to be executed
-    function applyMsg(SubnetID memory arrivingFrom, CrossMsg memory crossMsg) internal {
+    function applyMsg(SubnetID memory arrivingFrom, IpcEnvelope memory crossMsg) internal {
         GatewayActorStorage storage s = LibGatewayActorStorage.appStorage();
 
-        if (crossMsg.message.to.subnetId.isEmpty()) {
-            revert InvalidCrossMsgDstSubnet();
-        }
-
-        // If the crossnet destination is NOT the current network (network where the gateway is running),
-        // we add it to the postbox for further propagation.
-        if (!crossMsg.message.to.subnetId.equals(s.networkName)) {
-            bytes32 cid = crossMsg.toHash();
-            s.postbox[cid] = crossMsg;
+        if (crossMsg.to.subnetId.isEmpty()) {
+            sendReceipt(crossMsg, false, abi.encodeWithSelector(InvalidCrossMsgDstSubnet.selector));
```

Comment:
We need to distinguish between system-generated failures, and callee-generated failures. Currently it would be possible for a callee to trick the caller to think that the system generated a failure, if it throws a system error e.g. `InvalidCrossMsgDstSubnet`.

Proposal: make the result field an enum and have three values: Ok, SysErr, CallerErr,